src/librustc_data_structures/sip128.rs

```diff
@@ -122,42 +121,55 @@ impl SipHasher128 {
         self.state.v1 ^= 0xee;
     }
 
-    // Specialized write function that is only valid for buffers with len <= 8.
-    // It's used to force inlining of write_u8 and write_usize, those would normally be inlined
-    // except for composite types (that includes slices and str hashing because of delimiter).
-    // Without this extra push the compiler is very reluctant to inline delimiter writes,
-    // degrading performance substantially for the most common use cases.
+    // Specialized write function for values with size <= 8.
     #[inline]
-    fn short_write(&mut self, msg: &[u8]) {
-        debug_assert!(msg.len() <= 8);
-        let length = msg.len();
-        self.length += length;
+    fn short_write<T>(&mut self, _x: T, x: u64) {
```

Comment:
Hm, I wonder if it makes sense to go about this a bit differently, and instead have the "zero extension" be a trait like Into (but defined here) which goes from any primitive integer type to `u64` via zero extension (which is what we do below, I believe).

One option which we could explore is to do something like `MaybeUninit::zeroed().as_mut_ptr().offset(8 - std::mem::size_of::<T>()).cast::<T>.write(x)` as a way to get the zero extension working.

I could imagine that passing just (pointer, 64 bit integer) instead of pointer, number, 64 bit integer would allow for more aggressive inlining and so forth. It could also help LLVM, in theory, if it lets it see that the upper bits are zero (but that seems unlikely).

Reply:
I'm not sure what you mean by "more aggressive inlining". These functions are inlined (Cachegrind's results confirm that for me), and given that, there is no `_x` to actually pass, right? And presumably the `mem::size_of::<T>()` will end up being constants.

Reply:
Hm, if they're already inlined, I guess that's true, though I suspect that might be fairly brittle over time. I am not personally 100% clear on whether inlining is 'perfect' at erasing ABIs and such, too, but would suspect/hope that it is.

Feel free to resolve this.